File: gfapy/line/common/equivalence.py
import gfapy

class Equivalence:

  def __hash__(self):
    name = self.get("name")
    if name:
      return name.__hash__()
    else:
      return NotImplemented

  def __eq__(self, o):
    """
    Equivalence check

    Returns
    -------
    bool
      does the line has the same record type,
      contains the same tags
      and all positional fields and tags contain the same field values?

    See Also
    --------
    gfapy.line.edge.Link.__eq__
    """
    if o is self:
      return True
    if isinstance(o, str):
      name = self.get("name")
      if name:
        return name == str(o)
    if not isinstance(o, gfapy.Line):
      return False
    if (o.record_type != self.record_type):
      return False
    if sorted(o._data.keys()) != sorted(self._data.keys()):
      return False
    for k,v in o._data.items():
      if self._data[k] != v:
        if self.field_to_s(k) != o.field_to_s(k):
          return False
    return True

  def diff(self, other):
    if self.record_type != other.record_type:
      return ("incompatible", "record_type", \
              self.record_type, other.record_type)
    if self.__class__ != other.__class__:
      if self.version == other.version:
        raise gfapy.AssertionError
      return ("incompatible", "version", self.version, other.version)
    differences = []
    for fieldname in self.positional_fieldnames:
      value1 = self.field_to_s(fieldname)
      value2 = other.field_to_s(fieldname)
      if value1 != value2:
        differences.append(("different", "positional_field",
                            fieldname, value1, value2))
    for tagname in other.tagnames:
      if tagname not in self.tagnames:
        differences.append(("exclusive", ">", "tag",
                      tagname, other.get_datatype(tagname),
                      other.get(tagname)))
    for tagname in self.tagnames:
      if tagname not in other.tagnames:
        differences.append(("exclusive", "<", "tag",
                      tagname, self.get_datatype(tagname), self.get(tagname)))
      else:
        tag1 = self.field_to_s(tagname, tag=True)
        tag2 = other.field_to_s(tagname, tag=True)
        if tag1 != tag2:
          differences.append(("different", "tag", tagname,
                          self.get_datatype(tagname),
                          self.field_to_s(tagname),
                          other.get_datatype(tagname),
                          other.field_to_s(tagname)))
    return differences

  def diffscript(self, other, selfvar):
    outscript = []
    for diffitem in self.diff(other):
      if diffitem[0] == "incompatible":
        if diffitem[1] == "record_type":
          raise gfapy.RuntimeError(
            "Cannot compute conversion script: different record type\n"+
            "Line: {}\n".format(self)+
            "Other: {}\n".format(other)+
            "{0} != {1}",format(diffitem[2], diffitem[3]))
        elif diffitem[1] == "version":
          raise gfapy.RuntimeError(
            "Cannot compute conversion script: different GFA version\n"+
            "Line: {}\n".format(self)+
            "Other: {}\n".format(other)+
            "{0} != {1}",format(diffitem[2], diffitem[3]))
      elif diffitem[0] == "different":
        if diffitem[1] == "positional_field":
          outscript.append("{0}.set('{1}', '{2}')".format(selfvar,
                                      diffitem[2].replace("'","\\'"),
                                      diffitem[4].replace("'","\\'")))
        elif diffitem[1] == "tag":
          if diffitem[3] != diffitem[5]:
            outscript.append("{0}.set_datatype('{1}', '{2}')".format(selfvar,
                                      diffitem[2].replace("'","\\'"),
                                      diffitem[5].replace("'","\\'")))
          if diffitem[4] != diffitem[6]:
            outscript.append("{0}.set('{1}', '{2}')".format(selfvar,
                                      diffitem[2].replace("'","\\'"),
                                      diffitem[6].replace("'","\\'")))
      elif diffitem[0] == "exclusive":
        if diffitem[1] == ">":
          if diffitem[2] == "tag":
            outscript.append("{0}.set_datatype('{1}', '{2}')".format(selfvar,
                                      diffitem[3].replace("'","\\'"),
                                      diffitem[4].replace("'","\\'")))
            outscript.append("{0}.set('{1}', '{2}')".format(selfvar,
                                      diffitem[3].replace("'","\\'"),
                                      diffitem[5].replace("'","\\'")))
        elif diffitem[1] == "<":
          if diffitem[2] == "tag":
            outscript.append("{0}.delete('{1}')".format(selfvar,
                                      diffitem[3].replace("'","\\'")))
    return "\n".join(outscript)


  def _has_field_values(self, hsh, ignore_fields = None):
    assert(isinstance(hsh, dict))
    if ignore_fields is None:
      ignore_fields = []
    if ("record_type" in hsh) and ("record_type" not in ignore_fields) \
        and (self.record_type != hsh["record_type"]):
      return False
    ignore_fields.append("record_type")
    fieldnames = [i for i in hsh.keys() if i not in ignore_fields]
    for fieldname in fieldnames:
      value = self.get(fieldname)
      if value is None:
        return False
      if gfapy.is_placeholder(value):
        continue
      if value != hsh[fieldname] and \
          (self.field_to_s(fieldname) != hsh[fieldname]):
        return False
    return True

  def _has_eql_fields(self, refline, ignore_fields = None):
    assert(isinstance(refline, gfapy.Line))
    if ignore_fields is None:
      ignore_fields = []
    self._dealias_fieldnames(ignore_fields)
    if ("record_type" not in ignore_fields) \
        and (self.record_type != refline.record_type):
      return False
    fieldnames = refline.positional_fieldnames + refline.tagnames
    fieldnames = [i for i in fieldnames if i not in ignore_fields]
    if "name" in ignore_fields:
      name_field = refline.__class__.NAME_FIELD
      if name_field in fieldnames:
        fieldnames.remove(name_field)
    for fieldname in fieldnames:
      refvalue = refline.get(fieldname)
      if gfapy.is_placeholder(refvalue):
        continue
      value = self.get(fieldname)
      if value is None:
        return False
      if gfapy.is_placeholder(value):
        continue
      if value != refvalue:
        return False
    return True

File: lib/rgfa/line/common/equivalence.rb
# @tested_in unit_line_equivalence
module RGFA::Line::Common::Equivalence

  # Equivalence check
  # @param other [RGFA::Line, Symbol]
  # @return [Boolean] if other is a symbol, is it the same as
  #   the result of applying to_sym to the line?; if other is a line,
  #   does it have the same record type, positional fields and tags (value
  #   and datatype) as the line?
  # @see RGFA::Line::Edge::Link#==
  def ==(other)
    return self.to_sym == other.to_sym if other.kind_of?(Symbol)
    return false if !other.kind_of?(RGFA::Line)
    return false if (other.record_type != self.record_type)
    return false if other.data.keys.sort != data.keys.sort
    other.data.each do |k, v|
      if @data[k] != other.data[k]
        if field_to_s(k) != other.field_to_s(k)
          return false
        end
      end
    end
    return true
  end

  # Returns always false, as a line is not a placeholder (for compatibility
  # with other objects which can be representable as placeholders, such as
  # symbols, strings, arrays).
  # @return [false]
  def placeholder?
    false
  end

  # Computes the differences between the line and another line.
  # @return [Array<Array<Symbol, String>>] information about the differences;
  #   an empty array if no difference found
  def diff(other)
    if self.record_type != other.record_type
      return [:incompatible, :record_type, self.record_type, other.record_type]
    end
    if self.class != other.class
      raise RGFA::AssertionError if self.version == other.version
      return [:incompatible, :version, self.version, other.version]
    end
    differences = []
    positional_fieldnames.each do |fieldname|
      value1 = field_to_s(fieldname)
      value2 = other.field_to_s(fieldname)
      if value1 != value2
        differences << [:different, :positional_field,
                        fieldname, value1, value2]
      end
    end
    (self.tagnames - other.tagnames).each do |tagname|
      differences << [:exclusive, :<, :tag,
                      tagname, get_datatype(tagname), get(tagname)]
    end
    (other.tagnames - self.tagnames).each do |tagname|
      differences << [:exclusive, :>, :tag,
                      tagname, other.get_datatype(tagname), other.get(tagname)]
    end
    (self.tagnames & other.tagnames).each do |tagname|
      tag1 = field_to_s(tagname, tag: true)
      tag2 = other.field_to_s(tagname, tag: true)
      if tag1 != tag2
        differences << [:different, :tag, tagname,
                        get_datatype(tagname), field_to_s(tagname),
                        other.get_datatype(tagname), other.field_to_s(tagname)]
      end
    end
    return differences
  end

  # Computes a RGFA Ruby script for converting line into other
  # @return [String]
  def diffscript(other, selfvar)
    diffinfo = diff(other)
    outscript = []
    diffinfo.each do |diffitem|
      if diffitem[0] == :incompatible
        if diffitem[1] == :record_type
          raise RGFA::RuntimeError,
            "Cannot compute conversion script: different record type\n"+
            "Line: #{self}\n"+
            "Other: #{other}\n"+
            "#{diffitem[2]} != #{diffitem[3]}"
        elsif diffitem[1] == :version
          raise RGFA::RuntimeError,
            "Cannot compute conversion script: different GFA version\n"+
            "Line: #{self}\n"+
            "Other: #{other}\n"+
            "#{diffitem[2]} != #{diffitem[3]}"
        end
      elsif diffitem[0] == :different
        if diffitem[1] == :positional_field
          outscript <<
            "#{selfvar}.set(:#{diffitem[2]},'#{diffitem[4].gsub("'","\'")}')"
        elsif diffitem[1] == :tag
          if diffitem[3] != diffitem[5]
            outscript <<
              "#{selfvar}.set_datatype(:#{diffitem[2]},:#{diffitem[5]})"
          end
          if diffitem[4] != diffitem[6]
            outscript <<
              "#{selfvar}.set(:#{diffitem[2]},'#{diffitem[6].gsub("'","\'")}')"
          end
        end
      elsif diffitem[0] == :exclusive
        if diffitem[1] == :>
          if diffitem[2] == :tag
            outscript <<
              "#{selfvar}.set_datatype(:#{diffitem[3]},:#{diffitem[4]})"
            outscript <<
              "#{selfvar}.set(:#{diffitem[3]},'#{diffitem[5].gsub("'","\'")}')"
          end
        elsif diffitem[1] == :<
          if diffitem[2] == :tag
            outscript <<
              "#{selfvar}.delete(:#{diffitem[3]})"
          end
        end
      end
    end
    return outscript.join("\n")
  end

  # @api private
  module API_PRIVATE

    # Compares the field values line instance to an hash of
    # fieldnames => values.
    #
    # Each field in the line must have a value equal to that indicated
    # in the hash, except those indicated in the +ignore_fields+ array
    # and those containing placeholder values. The values can be decoded
    # (e.g. 1, an Integer) or encoded (e.g. "1", a String).
    #
    # @param hash [Hash(Symbol=>Object)] an hash value of fieldnames => values
    # @param ignore_fields [Array] list of hash keys to skip in the comparison
    #
    # @return [Boolean]
    def field_values?(hash, ignore_fields = [])
      if hash[:record_type] and !ignore_fields.include?(:record_type)
        return false if record_type != hash[:record_type]
      end
      ((hash.keys - ignore_fields) - [:record_type]).each do |fieldname|
        value = get(fieldname)
        return false if value.nil?
        next if value.placeholder?
        if value != hash[fieldname]
          return false if field_to_s(fieldname) != hash[fieldname]
        end
      end
      return true
    end

    # Compares the fields of the line to those of a reference line.
    #
    # The record type, positional fields and tags are compared, except
    # those listed in the +ignore_fields+ list. Fields containing placeholder
    # values in any of the two lines are ignored.
    #
    # @param refline [RGFA::Line] a reference line instance for the comparison
    # @param ignore_fields [Array] list of fields (record_type, positional
    #   fields and/or tags) to skip in the comparison; the special value +:name+
    #   will exclude the name field from the comparison
    #
    # @return [Boolean]
    def eql_fields?(refline, ignore_fields = [])
      dealias_fieldnames!(ignore_fields)
      unless ignore_fields.include?(:record_type)
        return false if self.record_type != refline.record_type
      end
      fieldnames = (refline.positional_fieldnames + refline.tagnames)
      fieldnames -= ignore_fields
      if ignore_fields.include?(:name)
        fieldnames.delete(refline.class::NAME_FIELD)
      end
      fieldnames.each do |fieldname|
        refvalue = refline.get(fieldname)
        next if refvalue.placeholder?
        value = get(fieldname)
        return false if value.nil?
        next if value.placeholder?
        return false if value != refvalue
      end
      return true
    end

  end
  include API_PRIVATE

end
